src/include/sof/drivers/ipc.h

```diff
@@ -116,6 +123,9 @@ struct ipc {
 	/* processing task */
 	struct task ipc_task;
 
+	/* FW callback on msg send */
+	struct ipc_send_event ipc_cb[SOF_CTRL_EVENT_COUNT];
```

Comment:
What if I queue 2 ipcs of the same type with different callbacks?

Reply:
The implementation has been change slightly. Now we have events for each IPC header (we treat header as some sort of ID). As for your question, I don't see a case where we send two exact same  IPCs at once. If that is the case for whatever reason then callback will be overwritten.